source/common/runtime/runtime_impl.cc

```diff
@@ -233,10 +233,53 @@ SnapshotImpl::Entry SnapshotImpl::createEntry(const std::string& value) {
   return entry;
 }
 
+void parseFractionValue(SnapshotImpl::Entry& entry, const ProtobufWkt::Struct& value) {
+  envoy::type::v3::FractionalPercent percent;
+  percent.set_denominator(envoy::type::v3::FractionalPercent::HUNDRED);
+  for (const auto& f : value.fields()) {
+    if (f.first == "numerator") {
+      if (f.second.has_number_value()) {
+        percent.set_numerator(f.second.number_value());
+      }
+    } else if (f.first == "denominator" && f.second.has_string_value()) {
+      if (f.second.string_value() == "HUNDRED") {
+        percent.set_denominator(envoy::type::v3::FractionalPercent::HUNDRED);
+      } else if (f.second.string_value() == "TEN_THOUSAND") {
+        percent.set_denominator(envoy::type::v3::FractionalPercent::TEN_THOUSAND);
+      } else if (f.second.string_value() == "MILLION") {
+        percent.set_denominator(envoy::type::v3::FractionalPercent::MILLION);
+      } else {
+        return;
+      }
+    } else {
+      return;
+    }
+  }
+
+  entry.fractional_percent_value_ = percent;
+}
+
 SnapshotImpl::Entry SnapshotImpl::createEntry(const ProtobufWkt::Value& value) {
-  // This isn't the smartest way to do it; we're round-tripping via YAML, this should be optimized
-  // if runtime parsing becomes performance sensitive.
-  return createEntry(MessageUtil::getYamlStringFromMessage(value, false, false));
+  Entry entry;
```

Comment:
Can we just make this `getJsonStringFromMessage` instead? Rationale is:
* This parsing and conversion is all fiddly and I'd probably want to see a bunch of unit tests, I don't think it's worth writing them and maintaining it unless we suspect performance really matters.
* In terms of getting rid of YAML dependency, we can use JSON here instead, since JSON is always a valid YAML subset, so the string parsing should still work.

Reply:
Sorry, I missed the bit in the linked issue referencing JSON. Can we at least factor these utlls into a library in common/common and add tests then?

Reply:
we can, but they're only used by runtime, and covered by the existing runtime unit tests, in particular https://github.com/envoyproxy/envoy/blob/main/test/common/runtime/runtime_impl_test.cc#L582 which I used to make sure invalid input was still handled the same way it was previously.

Reply:
actually swapping reviews as htuch mentioned he was ooo